example/advection/parthenon_app_inputs.cpp

```diff
@@ -209,12 +209,12 @@ Packages_t ProcessPackages(std::unique_ptr<ParameterInput> &pin) {
   Packages_t packages;
   auto pkg = advection_package::Initialize(pin.get());
   packages[pkg->label()] = pkg;
-  return packages;
-}
 
-void SetFillDerivedFunctions() {
-  parthenon::FillDerivedVariables::SetFillDerivedFunctions(advection_package::PreFill,
-                                                           advection_package::PostFill);
+  auto &app = packages["Parthenon::AppInput"];
+  app->AddParam("PreFillDerivedBlock", advection_package::PreFill);
+  app->AddParam("PostFillDerivedBlock", advection_package::PostFill);
```

Comment:
The type checking part, at least, is not so bad.  The `Params` class records the type of the thing being stashed and then checks it against the explicitly provided templated type of the `Get` call, throwing an error if they don't agree exactly.  I had this problem early on when using a `std::function` for the template parameter, which `Params` wouldn't allow.

As far as the other comments, I agree with them.  This is very flexible, and I like the fact that we're reusing the `Params` object in the `StateDescriptor` rather than having to include explicitly typed function pointers as member variables.  I agree good documentation is certainly a must.  The real disadvantage here in that regard is that you would never know there were these magic strings by reading the `StateDescriptor`/`Params` source.  Actually, perhaps that suggests an improvement -- there could be string literals defined in one of these headers that could be used everywhere of relevance, rather than just typing stuff out like I've done now.  Then there is at least _some_ measure of self-documentation in the source.  Thoughts?

Reply:
Replaced all the special but undocumented strings with constexpr char arrays.  I think I like this, including that the source is now more self documenting.

Reply:
I like the idea that the strings passed into `params` are themselves the documentation. That's a good idea.